### PR TITLE
[Core&Lib&Exmaples] Add balance method, its default algorithm

### PR DESCRIPTION
**Pull request by lmatz**
issue #130
Balance method can use different strategies to balance the number of objects in a particular object list among different threads.

How do I write the CMakeLists.txt under lib/?

### CODE REVIEW

**Comment by ddmbr** [comment]
I'm still reading but I think `lib/CMakeLists.txt` looks fine.

File: core/executor.hpp [hunk 1]
```diff
@@ -25,9 +25,64 @@
 #include "core/channel/channel_manager.hpp"
 #include "core/context.hpp"
 #include "core/objlist.hpp"
+#include "lib/balance_algo.hpp"
 
 namespace husky {
 
+// Algo is the implementation of the algorithm used to balance.
```
Comment (ddmbr) [inline core/executor.hpp]: You need to put things inside the `lib` namespace

**Comment by lmatz** [comment]
I forget to delete those lines added in the examples/CMakeLists.txt. I may modify it with other changes if any.

File: core/executor.hpp [hunk 2]
```diff
@@ -25,9 +25,64 @@
 #include "core/channel/channel_manager.hpp"
 #include "core/context.hpp"
 #include "core/objlist.hpp"
+#include "lib/balance_algo.hpp"
```
Comment (kygx-legend) [inline core/executor.hpp]: @ddmbr I think `core` should not depend on `lib`.

Comment (ddmbr) [inline core/executor.hpp]: Right. `core` should only include `core`.

Comment (lmatz) [inline core/executor.hpp]: So I should move lib/balance_algo.hpp to core/?

File: core/executor.hpp [hunk 3]
```diff
@@ -25,9 +25,64 @@
...
+    broadcast_channel.flush();
+
+    broadcast_channel.prepare_broadcast();
+    std::unordered_map<int,int> num_objs;
+    for(auto& global_tid : Context::get_global()->hash_ring.get_global_tids()) {
+        num_objs[global_tid] = broadcast_channel.get(global_tid);
+    }
+
+
+    // key: the global_tid of the thread which this current should send objects to
```
Comment (Yuzhen11) [inline core/executor.hpp]: `this current` -> `this current thread` ?

Comment (lmatz) [inline core/executor.hpp]: You are right, I will correct it.

**Comment by lmatz** [comment]
just curious what's the normal workflow of updating a pull request? Currently, I git commit --amend my new modifications and rebase on the 'master' if there is any new commits on branch 'master', then push to my 'dev' branch. Is that ok?

File: core/balance.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
+#include <queue>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "core/balance.hpp"
+#include "core/context.hpp"
```
Comment (kygx-legend) [inline core/balance.cpp]: It should be
```
#include "core/balance.hpp"

#include <queue>
#include <unordered_map>
#include <utility>
#include <vector>

#include "core/context.hpp"
```

File: core/balance.hpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <numeric>  // for std::accumulate
+#include <queue>
+#include <unordered_map>
+
+#include "core/context.hpp"
```
Comment (kygx-legend) [inline core/balance.hpp]: No need to include context.hpp in this file.

File: core/balance.hpp [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <numeric>  // for std::accumulate
```
Comment (kygx-legend) [inline core/balance.hpp]: Put this include to cpp since it is used there.

**Comment by kygx-legend** [comment]
@lmatz Precisely, it's to rebase your local branch to `upstream/master`. And push your local branch to `origin/dev` since this remote branch is connecting this PR.

**Comment by lmatz** [comment]
These mistakes should be checked manually by myself, right? It seems that lint.py and clang-format.py cannot output this kind of errors.

**Comment by kygx-legend** [comment]
Some of them. It can reduce the time to review if passing the scripts first. BTW, since github would not notice us about the latest change of the PR, please ping us when you think it's updated.

**Comment by lmatz** [comment]
I see. I will run script first and then check manually next time.

**Comment by lmatz** [comment]
forgot to say, it's been already updated according to the reviews last week. But "Changes requested" doesn't show the changes maybe because I did a forced push on my 'dev' branch.

**Comment by kygx-legend** [comment]
Could you add your information in `CREDITS` file?

**Comment by lmatz** [comment]
@kygx-legend  Sure, it's done.

**Comment by kygx-legend** [comment]
Great. Will merge then.

**Pull request merged**